### ISSUE TITLE
hyperkit: minikube start hangs at docker-machine `WaitForSSH` after host reboot

### ISSUE THREAD

**Issue by andsens** [kubernetes/minikube#3542]
**BUG REPORT**

**Environment**:

**Minikube version**: v0.32.0
- **OS**: macOS 10.14.2 (18C54)
- **VM Driver**: hyperkit
- **ISO version**: minikube-v0.32.0.iso
- **Install tools**: homebrew

**What happened**:

minikube hangs at "Starting VM", specifically the "wait for SSH" part (`About to run SSH command: exit 0`)

**What you expected to happen**:

minikube should complete the startup

**How to reproduce it**:

* start minikube and let it create the VM
* stop minikube
* reboot the host machine
* start minikube

**Output of `minikube logs` (if applicable)**:

```
-- Logs begin at Fri 2019-01-18 13:15:30 UTC, end at Fri 2019-01-18 13:20:27 UTC. --
-- No entries --
```

**Anything else do we need to know**:

I have been trying to pinpoint where the exact issue lies and seem to end up at docker-machine. However, docker-machine runs fine when invoked manually.

Here's the startup log for minikube:

```
minikube start -v=100
Aliases:
map[string]string{}
Override:
map[string]interface {}{"v":"100"}
PFlags:
map[string]viper.FlagValue{"iso-url":viper.pflagValue{flag:(*pflag.Flag)(0xc000327b80)}, "kubernetes-version":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4a00)}, "container-runtime":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b48c0)}, "disk-size":viper.pflagValue{flag:(*pflag.Flag)(0xc000327e00)}, "host-only-cidr":viper.pflagValue{flag:(*pflag.Flag)(0xc000327ea0)}, "hyperv-virtual-switch":viper.pflagValue{flag:(*pflag.Flag)(0xc000327f40)}, "memory":viper.pflagValue{flag:(*pflag.Flag)(0xc000327cc0)}, "mount-string":viper.pflagValue{flag:(*pflag.Flag)(0xc000327a40)}, "xhyve-disk-driver":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b40a0)}, "bootstrapper":viper.pflagValue{flag:(*pflag.Flag)(0xc000326fa0)}, "dns-domain":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4640)}, "enable-default-cni":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4b40)}, "apiserver-port":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b43c0)}, "kvm-network":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4000)}, "network-plugin":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4aa0)}, "docker-opt":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4320)}, "nfs-shares-root":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b41e0)}, "vm-driver":viper.pflagValue{flag:(*pflag.Flag)(0xc000327c20)}, "hyperkit-vsock-ports":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4f00)}, "mount":viper.pflagValue{flag:(*pflag.Flag)(0xc0003279a0)}, "apiserver-name":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4460)}, "apiserver-names":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4500)}, "cache-images":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4c80)}, "uuid":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4dc0)}, "docker-env":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4280)}, "extra-config":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4d20)}, "registry-mirror":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4820)}, "service-cluster-ip-range":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b46e0)}, "apiserver-ips":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b45a0)}, "gpu":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4fa0)}, "keep-context":viper.pflagValue{flag:(*pflag.Flag)(0xc000327900)}, "disable-driver-mounts":viper.pflagValue{flag:(*pflag.Flag)(0xc000327ae0)}, "feature-gates":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4be0)}, "hyperkit-vpnkit-sock":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4e60)}, "insecure-registry":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4780)}, "nfs-share":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4140)}, "profile":viper.pflagValue{flag:(*pflag.Flag)(0xc000326f00)}, "cpus":viper.pflagValue{flag:(*pflag.Flag)(0xc000327d60)}, "cri-socket":viper.pflagValue{flag:(*pflag.Flag)(0xc0001b4960)}}
Env:
map[string]string{}
Key/Value Store:
map[string]interface {}{}
Config:
map[string]interface {}{"wantreporterror":true, "cpus":2, "dashboard":false, "memory":5120, "vm-driver":"hyperkit"}
Defaults:
map[string]interface {}{"wantupdatenotification":true, "wantkubectldownloadmsg":true, "showdriverdeprecationnotification":true, "showbootstrapperdeprecationnotification":true, "log_dir":"", "reminderwaitperiodinhours":24, "wantreporterror":false, "wantreporterrorprompt":true, "wantnonedriverwarning":true, "v":"0", "alsologtostderr":"false"}
Starting local Kubernetes v1.13.2 cluster...
Starting VM...
Found binary path at /usr/local/bin/docker-machine-driver-hyperkit
Launching plugin server for driver hyperkit
Plugin server listening at address 127.0.0.1:52270
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(minikube) Calling .DriverName
(minikube) Calling .GetState
(minikube) Calling .Start
(minikube) Using UUID c0f82174-1a51-11e9-8f11-a8206629896a
(minikube) Generated MAC d2:f7:6:3d:fb:86
(minikube) Starting with cmdline: loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 systemd.legacy_systemd_cgroup_controller=yes base host=minikube
(minikube) Calling .GetConfigRaw
(minikube) Calling .DriverName
Waiting for SSH to be available...
Getting to WaitForSSH function...
(minikube) Calling .GetSSHHostname
(minikube) Calling .GetSSHPort
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x143c5c0] 0x143c590  [] 0s} 192.168.64.16 22 <nil> <nil>}
About to run SSH command:
exit 0
```

The process just hangs at `exit 0`, I've waited up to 1 hour sometimes, so it doesn't seem that any work is being performed.

Running docker-machine (version 0.16.1, build cce350d) while the process hangs yields the following:

```
$ docker-machine -D -s ~/.minikube --native-ssh ssh minikube
Docker Machine Version:  0.16.1, build cce350d
Found binary path at /usr/local/bin/docker-machine-driver-hyperkit
Launching plugin server for driver hyperkit
Plugin server listening at address 127.0.0.1:52277
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(minikube) Calling .GetState
(minikube) Calling .GetSSHHostname
(minikube) Calling .GetSSHPort
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x137ec60] 0x137ec30  [] 0s} 192.168.64.16 22 <nil> <nil>}
                         _             _
            _         _ ( )           ( )
  ___ ___  (_)  ___  (_)| |/')  _   _ | |_      __
/' _ ` _ `\| |/' _ `\| || , <  ( ) ( )| '_`\  /'__`\
| ( ) ( ) || || ( ) || || |\`\ | (_) || |_) )(  ___/
(_) (_) (_)(_)(_) (_)(_)(_) (_)`\___/'(_,__/'`\____)

$
```

I'm using the native client here because that seems to be what docker-machine in minikube is using (see `Using SSH client type: native`).

When running the docker-machine command near simultaneously with running `minikube start` I can get it to hang as well, but it eventually connects once the machine is ready.

`sudo screen ~/.minikube/machines/minikube/tty` also gives me a nice console with a prompt, so I can't for the life of me figure out what would be tripping up docker-machine in minikube.

**Comment by andsens** [comment]
OK, so I have a really bad workaround for this, maybe it helps others and/or maybe it helps finding the issue: When minikube hangs at `exit 0`, open a new terminal tab and run `docker-machine -D -s ~/.minikube --native-ssh ssh minikube` to see when the machine is actually running, once you have a prompt, just run `minikube start` once more to get it to initialize the machine. That's it, minikube should now be running.  
Also: Do not close the original `minikube start` command, hyperkit is a child process of it and you kill the machine by terminating it.

**Comment by andsens** [comment]
So I performed a fresh install of `minikube`, `docker-machine`, and `docker-machine-driver-hyperkit` through homebrew. `~/.minikube`, `~/.kube`, and `~/.docker` were all removed.  
I recreated `~/.minikube/config/config.json` with the following content:
```
{
    "cpus": 2,
    "dashboard": false,
    "ingress": false,
    "memory": 4096,
    "vm-driver": "hyperkit"
}
```
The problem persists. So this is definitely not a one-off.  
After that I suspected that my `~/.ssh` config may have something to do with it, so i created a completely fake clean `$HOME` and then retried.
```
$ mkdir fakehome
$ cd fakehome
$ ln -s ~/.minikube .minikube
$ env - PATH=$PATH HOME=$PWD minikube start -v=100
```

To no avail... The startup process still hangs at
```
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x143c5c0] 0x143c590  [] 0s} 192.168.64.17 22 <nil> <nil>}
About to run SSH command:
exit 0
```

**Comment by mightyguava** [comment]
I also have this problem. Had a minikube running yesterday off a fresh install, but now I'm unable to connect. Recreating it doesn't work either:

Creating
```
minikube -v 100 start --vm-driver hyperkit
😄  minikube v1.0.0 on darwin (amd64)
🤹  Downloading Kubernetes v1.14.0 images in the background ...
🔥  Creating hyperkit VM (CPUs=2, Memory=2048MB, Disk=20000MB) ...
Found binary path at /usr/local/bin/docker-machine-driver-hyperkit
Launching plugin server for driver hyperkit
Plugin server listening at address 127.0.0.1:49635
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(minikube) Calling .GetMachineName
(minikube) Calling .DriverName
Reading certificate data from /Users/yunchiluo/.minikube/certs/ca.pem
Decoding PEM data...
Parsing certificate...
Reading certificate data from /Users/yunchiluo/.minikube/certs/cert.pem
Decoding PEM data...
Parsing certificate...
Running pre-create checks...
(minikube) Calling .PreCreateCheck
(minikube) Calling .GetConfigRaw
Creating machine...
(minikube) Calling .Create
(minikube) Downloading /Users/yunchiluo/.minikube/cache/boot2docker.iso from file:///Users/yunchiluo/.minikube/cache/iso/minikube-v1.0.0.iso...
(minikube) DBG | 2019/03/29 11:08:53 [INFO] Creating ssh key...
(minikube) DBG | 2019/03/29 11:08:53 [INFO] Creating raw disk image...
(minikube) DBG | Writing magic tar header
(minikube) DBG | Writing SSH key tar header
(minikube) DBG | Mounting boot2docker.iso
(minikube) DBG | executing: &{/usr/bin/hdiutil [hdiutil attach /Users/yunchiluo/.minikube/machines/minikube/boot2docker.iso -mountpoint /Users/yunchiluo/.minikube/machines/minikube/b2d-image] []  <nil> 0xc4200c4008 0xc4200c4010 [] <nil> <nil> <nil> <nil> <nil> false [] [] [] [] <nil> <nil>} attach /Users/yunchiluo/.minikube/machines/minikube/boot2docker.iso -mountpoint /Users/yunchiluo/.minikube/machines/minikube/b2d-image
(minikube) /dev/disk2          	                               	/Users/yunchiluo/.minikube/machines/minikube/b2d-image
(minikube) DBG | Extracting Kernel Options...
(minikube) DBG | Extracted Options "loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 systemd.legacy_systemd_cgroup_controller=yes base host=minikube"
(minikube) DBG | Extracting /Users/yunchiluo/.minikube/machines/minikube/b2d-image/boot/bzImage into /Users/yunchiluo/.minikube/machines/minikube/bzImage
(minikube) DBG | Extracting /Users/yunchiluo/.minikube/machines/minikube/b2d-image/boot/initrd into /Users/yunchiluo/.minikube/machines/minikube/initrd
(minikube) DBG | Unmounting boot2docker.iso
(minikube) DBG | executing: &{/usr/bin/hdiutil [hdiutil detach /Users/yunchiluo/.minikube/machines/minikube/b2d-image] []  <nil> 0xc4200c4008 0xc4200c4010 [] <nil> <nil> <nil> <nil> <nil> false [] [] [] [] <nil> <nil>} detach /Users/yunchiluo/.minikube/machines/minikube/b2d-image
(minikube) "disk2" ejected.
(minikube) Using UUID 91041d72-5234-11e9-a79d-acde48001122
(minikube) Generated MAC ee:73:84:4c:cb:3a
(minikube) Starting with cmdline: loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 systemd.legacy_systemd_cgroup_controller=yes base host=minikube
(minikube) Calling .GetConfigRaw
(minikube) Calling .DriverName
(minikube) Calling .DriverName
Waiting for machine to be running, this may take a few minutes...
(minikube) Calling .GetState
Detecting operating system of created instance...
Waiting for SSH to be available...
Getting to WaitForSSH function...
(minikube) Calling .GetSSHHostname
(minikube) Calling .GetSSHPort
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x143d0e0] 0x143d0b0  [] 0s} 192.168.64.5 22 <nil> <nil>}
About to run SSH command:
exit 0
Error dialing TCP: dial tcp 192.168.64.5:22: connect: operation timed out
Error dialing TCP: dial tcp 192.168.64.5:22: connect: operation timed out
```

Connecting
```
docker-machine -D -s ~/.minikube --native-ssh ssh minikube
Docker Machine Version:  0.16.1, build cce350d7
Found binary path at /usr/local/bin/docker-machine-driver-hyperkit
Launching plugin server for driver hyperkit
Plugin server listening at address 127.0.0.1:49687
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(minikube) Calling .GetState
(minikube) Calling .GetSSHHostname
(minikube) Calling .GetSSHPort
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x134f690] 0x134f660  [] 0s} 192.168.64.5 22 <nil> <nil>}
dial tcp 192.168.64.5:22: connect: operation timed out
```

I noticed that the machine doesn't actually exist, which is weird....

```
docker-machine ls
NAME   ACTIVE   DRIVER   STATE   URL   SWARM   DOCKER   ERRORS
```

**Comment by andsens** [comment]
@mightyguava this is not the same issue. Your machine isn't actually reachable, which seems to be caused by it not being started at all. This issue refers to the ssh reachability test simply hanging even though it should succeed.

**Comment by mightyguava** [comment]
Ah good point. My machine is definitely running since I can see the `hyperkit` instance in ps, but yeah, not reachable at all via ssh.

**Comment by tstromberg** [comment]
This issue appears to be a duplicate of #2994, so I will close this one so that we may centralize the content relating to the issue. If you feel that this issue is not in fact a duplicate, please feel free to re-open it. 

If you run into this, please upgrade to the latest hyperkit driver we provide:

`curl -LO https://storage.googleapis.com/minikube/releases/latest/docker-machine-driver-hyperkit && sudo install -o root -g wheel -m 4755 docker-machine-driver-hyperkit /usr/local/bin/`

Thank you opening this bug report, and for your patience!
